// @lingo.dev/_sdk/build/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } } function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; } var _class; var _class2;// src/index.ts
var _zod = require('zod'); var _zod2 = _interopRequireDefault(_zod);
var __spec = require('@lingo.dev/_spec');
var _cuid2 = require('@paralleldrive/cuid2');
var engineParamsSchema = _zod2.default.object({
  apiKey: _zod2.default.string(),
  apiUrl: _zod2.default.string().url().default("https://engine.lingo.dev"),
  batchSize: _zod2.default.number().int().gt(0).lte(250).default(25),
  idealBatchItemSize: _zod2.default.number().int().gt(0).lte(2500).default(250)
}).passthrough();
var payloadSchema = _zod2.default.record(_zod2.default.string(), _zod2.default.any());
var referenceSchema = _zod2.default.record(__spec.localeCodeSchema, payloadSchema);
var hintsSchema = _zod2.default.record(_zod2.default.string(), _zod2.default.array(_zod2.default.string()));
var localizationParamsSchema = _zod2.default.object({
  sourceLocale: _zod2.default.union([__spec.localeCodeSchema, _zod2.default.null()]),
  targetLocale: __spec.localeCodeSchema,
  fast: _zod2.default.boolean().optional(),
  reference: referenceSchema.optional(),
  hints: hintsSchema.optional()
});
var LingoDotDevEngine = class {
  
  /**
   * Create a new LingoDotDevEngine instance
   * @param config - Configuration options for the Engine
   */
  constructor(config) {
    this.config = engineParamsSchema.parse(config);
  }
  /**
   * Localize content using the Lingo.dev API
   * @param payload - The content to be localized
   * @param params - Localization parameters including source/target locales and fast mode option
   * @param progressCallback - Optional callback function to report progress (0-100)
   * @param signal - Optional AbortSignal to cancel the operation
   * @returns Localized content
   * @internal
   */
  async _localizeRaw(payload, params, progressCallback, signal) {
    const finalPayload = payloadSchema.parse(payload);
    const finalParams = localizationParamsSchema.parse(params);
    const chunkedPayload = this.extractPayloadChunks(finalPayload);
    const processedPayloadChunks = [];
    const workflowId = _cuid2.createId.call(void 0, );
    for (let i = 0; i < chunkedPayload.length; i++) {
      const chunk = chunkedPayload[i];
      const percentageCompleted = Math.round(
        (i + 1) / chunkedPayload.length * 100
      );
      const processedPayloadChunk = await this.localizeChunk(
        finalParams.sourceLocale,
        finalParams.targetLocale,
        { data: chunk, reference: params.reference, hints: params.hints },
        workflowId,
        params.fast || false,
        signal
      );
      if (progressCallback) {
        progressCallback(percentageCompleted, chunk, processedPayloadChunk);
      }
      processedPayloadChunks.push(processedPayloadChunk);
    }
    return Object.assign({}, ...processedPayloadChunks);
  }
  /**
   * Localize a single chunk of content
   * @param sourceLocale - Source locale
   * @param targetLocale - Target locale
   * @param payload - Payload containing the chunk to be localized
   * @param workflowId - Workflow ID for tracking
   * @param fast - Whether to use fast mode
   * @param signal - Optional AbortSignal to cancel the operation
   * @returns Localized chunk
   */
  async localizeChunk(sourceLocale, targetLocale, payload, workflowId, fast, signal) {
    const res = await fetch(`${this.config.apiUrl}/i18n`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json; charset=utf-8",
        Authorization: `Bearer ${this.config.apiKey}`
      },
      body: JSON.stringify(
        {
          params: { workflowId, fast },
          locale: {
            source: sourceLocale,
            target: targetLocale
          },
          data: payload.data,
          reference: payload.reference,
          hints: payload.hints
        },
        null,
        2
      ),
      signal
    });
    if (!res.ok) {
      if (res.status >= 500 && res.status < 600) {
        const errorText = await res.text();
        throw new Error(
          `Server error (${res.status}): ${res.statusText}. ${errorText}. This may be due to temporary service issues.`
        );
      } else if (res.status === 400) {
        throw new Error(`Invalid request: ${res.statusText}`);
      } else {
        const errorText = await res.text();
        throw new Error(errorText);
      }
    }
    const jsonResponse = await res.json();
    if (!jsonResponse.data && jsonResponse.error) {
      throw new Error(jsonResponse.error);
    }
    return jsonResponse.data || {};
  }
  /**
   * Extract payload chunks based on the ideal chunk size
   * @param payload - The payload to be chunked
   * @returns An array of payload chunks
   */
  extractPayloadChunks(payload) {
    const result = [];
    let currentChunk = {};
    let currentChunkItemCount = 0;
    const payloadEntries = Object.entries(payload);
    for (let i = 0; i < payloadEntries.length; i++) {
      const [key, value] = payloadEntries[i];
      currentChunk[key] = value;
      currentChunkItemCount++;
      const currentChunkSize = this.countWordsInRecord(currentChunk);
      if (currentChunkSize > this.config.idealBatchItemSize || currentChunkItemCount >= this.config.batchSize || i === payloadEntries.length - 1) {
        result.push(currentChunk);
        currentChunk = {};
        currentChunkItemCount = 0;
      }
    }
    return result;
  }
  /**
   * Count words in a record or array
   * @param payload - The payload to count words in
   * @returns The total number of words
   */
  countWordsInRecord(payload) {
    if (Array.isArray(payload)) {
      return payload.reduce(
        (acc, item) => acc + this.countWordsInRecord(item),
        0
      );
    } else if (typeof payload === "object" && payload !== null) {
      return Object.values(payload).reduce(
        (acc, item) => acc + this.countWordsInRecord(item),
        0
      );
    } else if (typeof payload === "string") {
      return payload.trim().split(/\s+/).filter(Boolean).length;
    } else {
      return 0;
    }
  }
  /**
   * Localize a typical JavaScript object
   * @param obj - The object to be localized (strings will be extracted and translated)
   * @param params - Localization parameters:
   *   - sourceLocale: The source language code (e.g., 'en')
   *   - targetLocale: The target language code (e.g., 'es')
   *   - fast: Optional boolean to enable fast mode (faster but potentially lower quality)
   * @param progressCallback - Optional callback function to report progress (0-100)
   * @param signal - Optional AbortSignal to cancel the operation
   * @returns A new object with the same structure but localized string values
   */
  async localizeObject(obj, params, progressCallback, signal) {
    return this._localizeRaw(obj, params, progressCallback, signal);
  }
  /**
   * Localize a single text string
   * @param text - The text string to be localized
   * @param params - Localization parameters:
   *   - sourceLocale: The source language code (e.g., 'en')
   *   - targetLocale: The target language code (e.g., 'es')
   *   - fast: Optional boolean to enable fast mode (faster for bigger batches)
   * @param progressCallback - Optional callback function to report progress (0-100)
   * @param signal - Optional AbortSignal to cancel the operation
   * @returns The localized text string
   */
  async localizeText(text, params, progressCallback, signal) {
    const response = await this._localizeRaw(
      { text },
      params,
      progressCallback,
      signal
    );
    return response.text || "";
  }
  /**
   * Localize a text string to multiple target locales
   * @param text - The text string to be localized
   * @param params - Localization parameters:
   *   - sourceLocale: The source language code (e.g., 'en')
   *   - targetLocales: An array of target language codes (e.g., ['es', 'fr'])
   *   - fast: Optional boolean to enable fast mode (for bigger batches)
   * @param signal - Optional AbortSignal to cancel the operation
   * @returns An array of localized text strings
   */
  async batchLocalizeText(text, params, signal) {
    const responses = await Promise.all(
      params.targetLocales.map(
        (targetLocale) => this.localizeText(
          text,
          {
            sourceLocale: params.sourceLocale,
            targetLocale,
            fast: params.fast
          },
          void 0,
          signal
        )
      )
    );
    return responses;
  }
  /**
   * Localize an array of strings
   * @param strings - An array of strings to be localized
   * @param params - Localization parameters:
   *   - sourceLocale: The source language code (e.g., 'en')
   *   - targetLocale: The target language code (e.g., 'es')
   *   - fast: Optional boolean to enable fast mode (faster for bigger batches)
   * @returns An array of localized strings in the same order
   */
  async localizeStringArray(strings, params) {
    const mapped = strings.reduce(
      (acc, str, i) => {
        acc[`item_${i}`] = str;
        return acc;
      },
      {}
    );
    const result = await this.localizeObject(mapped, params);
    return Object.values(result);
  }
  /**
   * Localize a chat sequence while preserving speaker names
   * @param chat - Array of chat messages, each with 'name' and 'text' properties
   * @param params - Localization parameters:
   *   - sourceLocale: The source language code (e.g., 'en')
   *   - targetLocale: The target language code (e.g., 'es')
   *   - fast: Optional boolean to enable fast mode (faster but potentially lower quality)
   * @param progressCallback - Optional callback function to report progress (0-100)
   * @param signal - Optional AbortSignal to cancel the operation
   * @returns Array of localized chat messages with preserved structure
   */
  async localizeChat(chat, params, progressCallback, signal) {
    const localized = await this._localizeRaw(
      { chat },
      params,
      progressCallback,
      signal
    );
    return Object.entries(localized).map(([key, value]) => ({
      name: chat[parseInt(key.split("_")[1])].name,
      text: value
    }));
  }
  /**
   * Localize an HTML document while preserving structure and formatting
   * Handles both text content and localizable attributes (alt, title, placeholder, meta content)
   * @param html - The HTML document string to be localized
   * @param params - Localization parameters:
   *   - sourceLocale: The source language code (e.g., 'en')
   *   - targetLocale: The target language code (e.g., 'es')
   *   - fast: Optional boolean to enable fast mode (faster but potentially lower quality)
   * @param progressCallback - Optional callback function to report progress (0-100)
   * @param signal - Optional AbortSignal to cancel the operation
   * @returns The localized HTML document as a string, with updated lang attribute
   */
  async localizeHtml(html, params, progressCallback, signal) {
    const jsdomPackage = await Promise.resolve().then(() => _interopRequireWildcard(require("jsdom")));
    const { JSDOM } = jsdomPackage;
    const dom = new JSDOM(html);
    const document = dom.window.document;
    const LOCALIZABLE_ATTRIBUTES = {
      meta: ["content"],
      img: ["alt"],
      input: ["placeholder"],
      a: ["title"]
    };
    const UNLOCALIZABLE_TAGS = ["script", "style"];
    const extractedContent = {};
    const getPath = (node, attribute) => {
      const indices = [];
      let current = node;
      let rootParent = "";
      while (current) {
        const parent = current.parentElement;
        if (!parent) break;
        if (parent === document.documentElement) {
          rootParent = current.nodeName.toLowerCase();
          break;
        }
        const siblings = Array.from(parent.childNodes).filter(
          (n) => n.nodeType === 1 || n.nodeType === 3 && _optionalChain([n, 'access', _ => _.textContent, 'optionalAccess', _2 => _2.trim, 'call', _3 => _3()])
        );
        const index = siblings.indexOf(current);
        if (index !== -1) {
          indices.unshift(index);
        }
        current = parent;
      }
      const basePath = rootParent ? `${rootParent}/${indices.join("/")}` : indices.join("/");
      return attribute ? `${basePath}#${attribute}` : basePath;
    };
    const processNode = (node) => {
      let parent = node.parentElement;
      while (parent) {
        if (UNLOCALIZABLE_TAGS.includes(parent.tagName.toLowerCase())) {
          return;
        }
        parent = parent.parentElement;
      }
      if (node.nodeType === 3) {
        const text = _optionalChain([node, 'access', _4 => _4.textContent, 'optionalAccess', _5 => _5.trim, 'call', _6 => _6()]) || "";
        if (text) {
          extractedContent[getPath(node)] = text;
        }
      } else if (node.nodeType === 1) {
        const element = node;
        const tagName = element.tagName.toLowerCase();
        const attributes = LOCALIZABLE_ATTRIBUTES[tagName] || [];
        attributes.forEach((attr) => {
          const value = element.getAttribute(attr);
          if (value) {
            extractedContent[getPath(element, attr)] = value;
          }
        });
        Array.from(element.childNodes).filter(
          (n) => n.nodeType === 1 || n.nodeType === 3 && _optionalChain([n, 'access', _7 => _7.textContent, 'optionalAccess', _8 => _8.trim, 'call', _9 => _9()])
        ).forEach(processNode);
      }
    };
    Array.from(document.head.childNodes).filter(
      (n) => n.nodeType === 1 || n.nodeType === 3 && _optionalChain([n, 'access', _10 => _10.textContent, 'optionalAccess', _11 => _11.trim, 'call', _12 => _12()])
    ).forEach(processNode);
    Array.from(document.body.childNodes).filter(
      (n) => n.nodeType === 1 || n.nodeType === 3 && _optionalChain([n, 'access', _13 => _13.textContent, 'optionalAccess', _14 => _14.trim, 'call', _15 => _15()])
    ).forEach(processNode);
    const localizedContent = await this._localizeRaw(
      extractedContent,
      params,
      progressCallback,
      signal
    );
    document.documentElement.setAttribute("lang", params.targetLocale);
    Object.entries(localizedContent).forEach(([path, value]) => {
      const [nodePath, attribute] = path.split("#");
      const [rootTag, ...indices] = nodePath.split("/");
      let parent = rootTag === "head" ? document.head : document.body;
      let current = parent;
      for (const index of indices) {
        const siblings = Array.from(parent.childNodes).filter(
          (n) => n.nodeType === 1 || n.nodeType === 3 && _optionalChain([n, 'access', _16 => _16.textContent, 'optionalAccess', _17 => _17.trim, 'call', _18 => _18()])
        );
        current = siblings[parseInt(index)] || null;
        if (_optionalChain([current, 'optionalAccess', _19 => _19.nodeType]) === 1) {
          parent = current;
        }
      }
      if (current) {
        if (attribute) {
          current.setAttribute(attribute, value);
        } else {
          current.textContent = value;
        }
      }
    });
    return dom.serialize();
  }
  /**
   * Detect the language of a given text
   * @param text - The text to analyze
   * @param signal - Optional AbortSignal to cancel the operation
   * @returns Promise resolving to a locale code (e.g., 'en', 'es', 'fr')
   */
  async recognizeLocale(text, signal) {
    const response = await fetch(`${this.config.apiUrl}/recognize`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json; charset=utf-8",
        Authorization: `Bearer ${this.config.apiKey}`
      },
      body: JSON.stringify({ text }),
      signal
    });
    if (!response.ok) {
      if (response.status >= 500 && response.status < 600) {
        throw new Error(
          `Server error (${response.status}): ${response.statusText}. This may be due to temporary service issues.`
        );
      }
      throw new Error(`Error recognizing locale: ${response.statusText}`);
    }
    const jsonResponse = await response.json();
    return jsonResponse.locale;
  }
  async whoami(signal) {
    try {
      const res = await fetch(`${this.config.apiUrl}/whoami`, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.config.apiKey}`,
          ContentType: "application/json"
        },
        signal
      });
      if (res.ok) {
        const payload = await res.json();
        if (!_optionalChain([payload, 'optionalAccess', _20 => _20.email])) {
          return null;
        }
        return {
          email: payload.email,
          id: payload.id
        };
      }
      if (res.status >= 500 && res.status < 600) {
        throw new Error(
          `Server error (${res.status}): ${res.statusText}. This may be due to temporary service issues.`
        );
      }
      return null;
    } catch (error) {
      if (error instanceof Error && error.message.includes("Server error")) {
        throw error;
      }
      return null;
    }
  }
};
var ReplexicaEngine = (_class = class _ReplexicaEngine extends LingoDotDevEngine {
  static __initStatic() {this.hasWarnedDeprecation = false}
  constructor(config) {
    super(config);
    if (!_ReplexicaEngine.hasWarnedDeprecation) {
      console.warn(
        "ReplexicaEngine is deprecated and will be removed in a future release. Please use LingoDotDevEngine instead. See https://lingo.dev/cli for more information."
      );
      _ReplexicaEngine.hasWarnedDeprecation = true;
    }
  }
}, _class.__initStatic(), _class);
var LingoEngine = (_class2 = class _LingoEngine extends LingoDotDevEngine {
  static __initStatic2() {this.hasWarnedDeprecation = false}
  constructor(config) {
    super(config);
    if (!_LingoEngine.hasWarnedDeprecation) {
      console.warn(
        "LingoEngine is deprecated and will be removed in a future release. Please use LingoDotDevEngine instead. See https://lingo.dev/cli for more information."
      );
      _LingoEngine.hasWarnedDeprecation = true;
    }
  }
}, _class2.__initStatic2(), _class2);




exports.LingoDotDevEngine = LingoDotDevEngine; exports.LingoEngine = LingoEngine; exports.ReplexicaEngine = ReplexicaEngine;
