// @lingo.dev/_locales/build/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  LOCALE_REGEX: () => LOCALE_REGEX,
  getCountryName: () => getCountryName,
  getLanguageCode: () => getLanguageCode,
  getLanguageName: () => getLanguageName,
  getRegionCode: () => getRegionCode,
  getScriptCode: () => getScriptCode,
  getScriptName: () => getScriptName,
  isValidLanguageCode: () => isValidLanguageCode,
  isValidLocale: () => isValidLocale,
  isValidRegionCode: () => isValidRegionCode,
  isValidScriptCode: () => isValidScriptCode,
  parseLocale: () => parseLocale,
  parseLocaleWithDetails: () => parseLocaleWithDetails
});
module.exports = __toCommonJS(src_exports);

// src/constants.ts
var LOCALE_REGEX = /^([a-z]{2,3})(?:[-_]([A-Za-z]{4}))?(?:[-_]([A-Z]{2}|[0-9]{3}))?$/;

// src/parser.ts
function normalizeLocaleCase(locale) {
  const parts = locale.split(/[-_]/);
  if (parts.length === 1) {
    return parts[0].toLowerCase();
  }
  if (parts.length === 2) {
    const language = parts[0].toLowerCase();
    const region = parts[1].toUpperCase();
    return `${language}-${region}`;
  }
  if (parts.length === 3) {
    const language = parts[0].toLowerCase();
    const script = parts[1];
    const region = parts[2].toUpperCase();
    return `${language}-${script}-${region}`;
  }
  return locale;
}
function parseLocale(locale) {
  if (typeof locale !== "string") {
    throw new Error("Locale must be a string");
  }
  if (!locale.trim()) {
    throw new Error("Locale cannot be empty");
  }
  const normalizedLocale = normalizeLocaleCase(locale);
  const match = normalizedLocale.match(LOCALE_REGEX);
  if (!match) {
    throw new Error(`Invalid locale format: ${locale}`);
  }
  const [, language, script, region] = match;
  const components = {
    language: language.toLowerCase()
  };
  if (script) {
    components.script = script;
  }
  if (region) {
    components.region = region.toUpperCase();
  }
  return components;
}
function parseLocaleWithDetails(locale) {
  try {
    const components = parseLocale(locale);
    let delimiter = null;
    if (locale.includes("-")) {
      delimiter = "-";
    } else if (locale.includes("_")) {
      delimiter = "_";
    }
    return {
      components,
      delimiter,
      isValid: true
    };
  } catch (error) {
    return {
      components: { language: "" },
      delimiter: null,
      isValid: false,
      error: error instanceof Error ? error.message : "Unknown parsing error"
    };
  }
}
function getLanguageCode(locale) {
  return parseLocale(locale).language;
}
function getScriptCode(locale) {
  const components = parseLocale(locale);
  return components.script || null;
}
function getRegionCode(locale) {
  const components = parseLocale(locale);
  return components.region || null;
}

// src/validation.ts
var VALID_LANGUAGE_CODES = /* @__PURE__ */ new Set([
  "aa",
  "ab",
  "ae",
  "af",
  "ak",
  "am",
  "an",
  "ar",
  "as",
  "av",
  "ay",
  "az",
  "ba",
  "be",
  "bg",
  "bh",
  "bi",
  "bm",
  "bn",
  "bo",
  "br",
  "bs",
  "ca",
  "ce",
  "ch",
  "co",
  "cr",
  "cs",
  "cu",
  "cv",
  "cy",
  "da",
  "de",
  "dv",
  "dz",
  "ee",
  "el",
  "en",
  "eo",
  "es",
  "et",
  "eu",
  "fa",
  "ff",
  "fi",
  "fj",
  "fo",
  "fr",
  "fy",
  "ga",
  "gd",
  "gl",
  "gn",
  "gu",
  "gv",
  "ha",
  "he",
  "hi",
  "ho",
  "hr",
  "ht",
  "hu",
  "hy",
  "hz",
  "ia",
  "id",
  "ie",
  "ig",
  "ii",
  "ik",
  "io",
  "is",
  "it",
  "iu",
  "ja",
  "jv",
  "ka",
  "kg",
  "ki",
  "kj",
  "kk",
  "kl",
  "km",
  "kn",
  "ko",
  "kr",
  "ks",
  "ku",
  "kv",
  "kw",
  "ky",
  "la",
  "lb",
  "lg",
  "li",
  "ln",
  "lo",
  "lt",
  "lu",
  "lv",
  "mg",
  "mh",
  "mi",
  "mk",
  "ml",
  "mn",
  "mr",
  "ms",
  "mt",
  "my",
  "na",
  "nb",
  "nd",
  "ne",
  "ng",
  "nl",
  "nn",
  "no",
  "nr",
  "nv",
  "ny",
  "oc",
  "oj",
  "om",
  "or",
  "os",
  "pa",
  "pi",
  "pl",
  "ps",
  "pt",
  "qu",
  "rm",
  "rn",
  "ro",
  "ru",
  "rw",
  "sa",
  "sc",
  "sd",
  "se",
  "sg",
  "si",
  "sk",
  "sl",
  "sm",
  "sn",
  "so",
  "sq",
  "sr",
  "ss",
  "st",
  "su",
  "sv",
  "sw",
  "ta",
  "te",
  "tg",
  "th",
  "ti",
  "tk",
  "tl",
  "tn",
  "to",
  "tr",
  "ts",
  "tt",
  "tw",
  "ty",
  "ug",
  "uk",
  "ur",
  "uz",
  "ve",
  "vi",
  "vo",
  "wa",
  "wo",
  "xh",
  "yi",
  "yo",
  "za",
  "zh",
  "zu"
]);
var VALID_SCRIPT_CODES = /* @__PURE__ */ new Set([
  "Adlm",
  "Afak",
  "Aghb",
  "Ahom",
  "Arab",
  "Aran",
  "Armi",
  "Armn",
  "Avst",
  "Bali",
  "Bamu",
  "Bass",
  "Batk",
  "Beng",
  "Bhks",
  "Blis",
  "Bopo",
  "Brah",
  "Brai",
  "Bugi",
  "Buhd",
  "Cakm",
  "Cans",
  "Cari",
  "Cham",
  "Cher",
  "Chrs",
  "Cirt",
  "Copt",
  "Cpmn",
  "Cprt",
  "Cyrl",
  "Cyrs",
  "Deva",
  "Diak",
  "Dogr",
  "Dsrt",
  "Dupl",
  "Egyd",
  "Egyh",
  "Egyp",
  "Elba",
  "Elym",
  "Ethi",
  "Gara",
  "Gong",
  "Gonm",
  "Goth",
  "Gran",
  "Grek",
  "Gujr",
  "Guru",
  "Hanb",
  "Hang",
  "Hani",
  "Hano",
  "Hans",
  "Hant",
  "Hatr",
  "Hebr",
  "Hira",
  "Hluw",
  "Hmng",
  "Hmnp",
  "Hrkt",
  "Hung",
  "Inds",
  "Ital",
  "Jamo",
  "Java",
  "Jpan",
  "Jurc",
  "Kali",
  "Kana",
  "Khar",
  "Khmr",
  "Khoj",
  "Kits",
  "Knda",
  "Kore",
  "Kpel",
  "Kthi",
  "Lana",
  "Laoo",
  "Latf",
  "Latg",
  "Latn",
  "Leke",
  "Lepc",
  "Limb",
  "Lina",
  "Linb",
  "Lisu",
  "Loma",
  "Lyci",
  "Lydi",
  "Mahj",
  "Maka",
  "Mand",
  "Mani",
  "Marc",
  "Maya",
  "Medf",
  "Mend",
  "Merc",
  "Mero",
  "Mlym",
  "Modi",
  "Mong",
  "Moon",
  "Mroo",
  "Mtei",
  "Mult",
  "Mymr",
  "Nand",
  "Narb",
  "Nbat",
  "Newa",
  "Nkgb",
  "Nkoo",
  "Nshu",
  "Ogam",
  "Olck",
  "Orkh",
  "Orya",
  "Osge",
  "Osma",
  "Ougr",
  "Palm",
  "Pauc",
  "Perm",
  "Phag",
  "Phli",
  "Phlp",
  "Phlv",
  "Phnx",
  "Plrd",
  "Prti",
  "Qaaa",
  "Qabx",
  "Rjng",
  "Rohg",
  "Roro",
  "Runr",
  "Samr",
  "Sara",
  "Sarb",
  "Saur",
  "Sgnw",
  "Shaw",
  "Shrd",
  "Shui",
  "Sidd",
  "Sind",
  "Sinh",
  "Sogd",
  "Sogo",
  "Sora",
  "Soyo",
  "Sund",
  "Sylo",
  "Syrc",
  "Syre",
  "Syrj",
  "Syrn",
  "Tagb",
  "Takr",
  "Tale",
  "Talu",
  "Taml",
  "Tang",
  "Tavt",
  "Telu",
  "Teng",
  "Tfng",
  "Tglg",
  "Thaa",
  "Thai",
  "Tibt",
  "Tirh",
  "Ugar",
  "Vaii",
  "Visp",
  "Wara",
  "Wcho",
  "Wole",
  "Xpeo",
  "Xsux",
  "Yezi",
  "Yiii",
  "Zanb",
  "Zinh",
  "Zmth",
  "Zsye",
  "Zsym",
  "Zxxx",
  "Zyyy",
  "Zzzz"
]);
var VALID_REGION_CODES = /* @__PURE__ */ new Set([
  "AD",
  "AE",
  "AF",
  "AG",
  "AI",
  "AL",
  "AM",
  "AO",
  "AQ",
  "AR",
  "AS",
  "AT",
  "AU",
  "AW",
  "AX",
  "AZ",
  "BA",
  "BB",
  "BD",
  "BE",
  "BF",
  "BG",
  "BH",
  "BI",
  "BJ",
  "BL",
  "BM",
  "BN",
  "BO",
  "BQ",
  "BR",
  "BS",
  "BT",
  "BV",
  "BW",
  "BY",
  "BZ",
  "CA",
  "CC",
  "CD",
  "CF",
  "CG",
  "CH",
  "CI",
  "CK",
  "CL",
  "CM",
  "CN",
  "CO",
  "CR",
  "CU",
  "CV",
  "CW",
  "CX",
  "CY",
  "CZ",
  "DE",
  "DJ",
  "DK",
  "DM",
  "DO",
  "DZ",
  "EC",
  "EE",
  "EG",
  "EH",
  "ER",
  "ES",
  "ET",
  "FI",
  "FJ",
  "FK",
  "FM",
  "FO",
  "FR",
  "GA",
  "GB",
  "GD",
  "GE",
  "GF",
  "GG",
  "GH",
  "GI",
  "GL",
  "GM",
  "GN",
  "GP",
  "GQ",
  "GR",
  "GS",
  "GT",
  "GU",
  "GW",
  "GY",
  "HK",
  "HM",
  "HN",
  "HR",
  "HT",
  "HU",
  "ID",
  "IE",
  "IL",
  "IM",
  "IN",
  "IO",
  "IQ",
  "IR",
  "IS",
  "IT",
  "JE",
  "JM",
  "JO",
  "JP",
  "KE",
  "KG",
  "KH",
  "KI",
  "KM",
  "KN",
  "KP",
  "KR",
  "KW",
  "KY",
  "KZ",
  "LA",
  "LB",
  "LC",
  "LI",
  "LK",
  "LR",
  "LS",
  "LT",
  "LU",
  "LV",
  "LY",
  "MA",
  "MC",
  "MD",
  "ME",
  "MF",
  "MG",
  "MH",
  "MK",
  "ML",
  "MM",
  "MN",
  "MO",
  "MP",
  "MQ",
  "MR",
  "MS",
  "MT",
  "MU",
  "MV",
  "MW",
  "MX",
  "MY",
  "MZ",
  "NA",
  "NC",
  "NE",
  "NF",
  "NG",
  "NI",
  "NL",
  "NO",
  "NP",
  "NR",
  "NU",
  "NZ",
  "OM",
  "PA",
  "PE",
  "PF",
  "PG",
  "PH",
  "PK",
  "PL",
  "PM",
  "PN",
  "PR",
  "PS",
  "PT",
  "PW",
  "PY",
  "QA",
  "RE",
  "RO",
  "RS",
  "RU",
  "RW",
  "SA",
  "SB",
  "SC",
  "SD",
  "SE",
  "SG",
  "SH",
  "SI",
  "SJ",
  "SK",
  "SL",
  "SM",
  "SN",
  "SO",
  "SR",
  "SS",
  "ST",
  "SV",
  "SX",
  "SY",
  "SZ",
  "TC",
  "TD",
  "TF",
  "TG",
  "TH",
  "TJ",
  "TK",
  "TL",
  "TM",
  "TN",
  "TO",
  "TR",
  "TT",
  "TV",
  "TW",
  "TZ",
  "UA",
  "UG",
  "UM",
  "US",
  "UY",
  "UZ",
  "VA",
  "VC",
  "VE",
  "VG",
  "VI",
  "VN",
  "VU",
  "WF",
  "WS",
  "YE",
  "YT",
  "ZA",
  "ZM",
  "ZW"
]);
var VALID_NUMERIC_REGION_CODES = /* @__PURE__ */ new Set([
  "001",
  "002",
  "003",
  "005",
  "009",
  "010",
  "011",
  "013",
  "014",
  "015",
  "017",
  "018",
  "019",
  "021",
  "029",
  "030",
  "034",
  "035",
  "039",
  "053",
  "054",
  "057",
  "061",
  "142",
  "143",
  "145",
  "150",
  "151",
  "154",
  "155",
  "202",
  "419",
  "AC",
  "BL",
  "BQ",
  "BV",
  "CP",
  "CW",
  "DG",
  "EA",
  "EU",
  "EZ",
  "FK",
  "FO",
  "GF",
  "GG",
  "GI",
  "GL",
  "GP",
  "GS",
  "GU",
  "HM",
  "IC",
  "IM",
  "IO",
  "JE",
  "KY",
  "MF",
  "MH",
  "MO",
  "MP",
  "MQ",
  "MS",
  "NC",
  "NF",
  "PF",
  "PM",
  "PN",
  "PR",
  "PS",
  "RE",
  "SH",
  "SJ",
  "SX",
  "TC",
  "TF",
  "TK",
  "TL",
  "UM",
  "VA",
  "VC",
  "VG",
  "VI",
  "WF",
  "YT"
]);
function isValidLocale(locale) {
  if (typeof locale !== "string" || !locale.trim()) {
    return false;
  }
  try {
    const match = locale.match(LOCALE_REGEX);
    if (!match) {
      return false;
    }
    const [, language, script, region] = match;
    if (!isValidLanguageCode(language)) {
      return false;
    }
    if (script && !isValidScriptCode(script)) {
      return false;
    }
    if (region && !isValidRegionCode(region)) {
      return false;
    }
    return true;
  } catch {
    return false;
  }
}
function isValidLanguageCode(code) {
  if (typeof code !== "string" || !code.trim()) {
    return false;
  }
  return VALID_LANGUAGE_CODES.has(code.toLowerCase());
}
function isValidScriptCode(code) {
  if (typeof code !== "string" || !code.trim()) {
    return false;
  }
  return VALID_SCRIPT_CODES.has(code);
}
function isValidRegionCode(code) {
  if (typeof code !== "string" || !code.trim()) {
    return false;
  }
  const upperCode = code.toUpperCase();
  return VALID_REGION_CODES.has(upperCode) || VALID_NUMERIC_REGION_CODES.has(upperCode);
}

// src/names/loader.ts
var CLDR_BASE_URL = process.env.CLDR_BASE_URL || "https://raw.githubusercontent.com/unicode-org/cldr-json/main/cldr-json/cldr-localenames-full/main";
var cache = /* @__PURE__ */ new Map();
async function loadTerritoryNames(displayLanguage) {
  const cacheKey = `territories-${displayLanguage}`;
  if (cache.has(cacheKey)) {
    return cache.get(cacheKey);
  }
  try {
    const response = await fetch(
      `${CLDR_BASE_URL}/${displayLanguage}/territories.json`
    );
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }
    const data = await response.json();
    const territories = data?.main?.[displayLanguage]?.localeDisplayNames?.territories || {};
    cache.set(cacheKey, territories);
    return territories;
  } catch (error) {
    if (displayLanguage !== "en") {
      console.warn(
        `Failed to load territory names for ${displayLanguage}, falling back to English`
      );
      return loadTerritoryNames("en");
    }
    throw new Error(
      `Failed to load territory names for ${displayLanguage}: ${error}`
    );
  }
}
async function loadLanguageNames(displayLanguage) {
  const cacheKey = `languages-${displayLanguage}`;
  if (cache.has(cacheKey)) {
    return cache.get(cacheKey);
  }
  try {
    const response = await fetch(
      `${CLDR_BASE_URL}/${displayLanguage}/languages.json`
    );
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }
    const data = await response.json();
    const languages = data?.main?.[displayLanguage]?.localeDisplayNames?.languages || {};
    cache.set(cacheKey, languages);
    return languages;
  } catch (error) {
    if (displayLanguage !== "en") {
      console.warn(
        `Failed to load language names for ${displayLanguage}, falling back to English`
      );
      return loadLanguageNames("en");
    }
    throw new Error(
      `Failed to load language names for ${displayLanguage}: ${error}`
    );
  }
}
async function loadScriptNames(displayLanguage) {
  const cacheKey = `scripts-${displayLanguage}`;
  if (cache.has(cacheKey)) {
    return cache.get(cacheKey);
  }
  try {
    const response = await fetch(
      `${CLDR_BASE_URL}/${displayLanguage}/scripts.json`
    );
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }
    const data = await response.json();
    const scripts = data?.main?.[displayLanguage]?.localeDisplayNames?.scripts || {};
    const enhancedScripts = { ...scripts };
    if (scripts["Hans-alt-stand-alone"]) {
      enhancedScripts.Hans = scripts["Hans-alt-stand-alone"];
    }
    if (scripts["Hant-alt-stand-alone"]) {
      enhancedScripts.Hant = scripts["Hant-alt-stand-alone"];
    }
    cache.set(cacheKey, enhancedScripts);
    return enhancedScripts;
  } catch (error) {
    if (displayLanguage !== "en") {
      console.warn(
        `Failed to load script names for ${displayLanguage}, falling back to English`
      );
      return loadScriptNames("en");
    }
    throw new Error(
      `Failed to load script names for ${displayLanguage}: ${error}`
    );
  }
}

// src/names/index.ts
async function getCountryName(countryCode, displayLanguage = "en") {
  if (!countryCode) {
    throw new Error("Country code is required");
  }
  const territories = await loadTerritoryNames(displayLanguage);
  const name = territories[countryCode.toUpperCase()];
  if (!name) {
    throw new Error(`Country code "${countryCode}" not found`);
  }
  return name;
}
async function getLanguageName(languageCode, displayLanguage = "en") {
  if (!languageCode) {
    throw new Error("Language code is required");
  }
  const languages = await loadLanguageNames(displayLanguage);
  const name = languages[languageCode.toLowerCase()];
  if (!name) {
    throw new Error(`Language code "${languageCode}" not found`);
  }
  return name;
}
async function getScriptName(scriptCode, displayLanguage = "en") {
  if (!scriptCode) {
    throw new Error("Script code is required");
  }
  const scripts = await loadScriptNames(displayLanguage);
  const name = scripts[scriptCode];
  if (!name) {
    throw new Error(`Script code "${scriptCode}" not found`);
  }
  return name;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LOCALE_REGEX,
  getCountryName,
  getLanguageCode,
  getLanguageName,
  getRegionCode,
  getScriptCode,
  getScriptName,
  isValidLanguageCode,
  isValidLocale,
  isValidRegionCode,
  isValidScriptCode,
  parseLocale,
  parseLocaleWithDetails
});
