// @lingo.dev/_compiler/build/lingo-turbopack-loader.mjs
import {
  loadDictionary,
  transformComponent
} from "./chunk-COKHINWO.mjs";

// src/lingo-turbopack-loader.ts
async function lingo_turbopack_loader_default(source) {
  const callback = this.async();
  const params = this.getOptions();
  const isDev = process.env.NODE_ENV !== "production";
  try {
    const dictionary = await loadDictionary({
      resourcePath: this.resourcePath,
      resourceQuery: this.resourceQuery,
      params,
      sourceRoot: params.sourceRoot,
      lingoDir: params.lingoDir,
      isDev
    });
    if (dictionary) {
      const code = `export default ${JSON.stringify(dictionary, null, 2)};`;
      return callback(null, code);
    }
    const result = transformComponent({
      code: source,
      params,
      resourcePath: this.resourcePath,
      sourceRoot: params.sourceRoot
    });
    return callback(
      null,
      result.code,
      result.map ? JSON.stringify(result.map) : void 0
    );
  } catch (error) {
    console.error(
      `\u26A0\uFE0F  Lingo.dev compiler (Turbopack) failed for ${this.resourcePath}:`
    );
    console.error("\u26A0\uFE0F  Details:", error);
    callback(error);
  }
}
export {
  lingo_turbopack_loader_default as default
};
