// @lingo.dev/_compiler/build/lingo-turbopack-loader.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _chunkTMPC4O5Vcjs = require('./chunk-TMPC4O5V.cjs');

// src/lingo-turbopack-loader.ts
async function lingo_turbopack_loader_default(source) {
  const callback = this.async();
  const params = this.getOptions();
  const isDev = process.env.NODE_ENV !== "production";
  try {
    const dictionary = await _chunkTMPC4O5Vcjs.loadDictionary.call(void 0, {
      resourcePath: this.resourcePath,
      resourceQuery: this.resourceQuery,
      params,
      sourceRoot: params.sourceRoot,
      lingoDir: params.lingoDir,
      isDev
    });
    if (dictionary) {
      const code = `export default ${JSON.stringify(dictionary, null, 2)};`;
      return callback(null, code);
    }
    const result = _chunkTMPC4O5Vcjs.transformComponent.call(void 0, {
      code: source,
      params,
      resourcePath: this.resourcePath,
      sourceRoot: params.sourceRoot
    });
    return callback(
      null,
      result.code,
      result.map ? JSON.stringify(result.map) : void 0
    );
  } catch (error) {
    console.error(
      `\u26A0\uFE0F  Lingo.dev compiler (Turbopack) failed for ${this.resourcePath}:`
    );
    console.error("\u26A0\uFE0F  Details:", error);
    callback(error);
  }
}


exports.default = lingo_turbopack_loader_default;

module.exports = exports.default;
