// @datocms/cma-client/dist/cjs/generated/resources/SchemaMenuItem.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Utils = __importStar(require("@datocms/rest-client-utils"));
var BaseResource_1 = __importDefault(require("../../BaseResource"));
var SchemaMenuItem = /** @class */ (function (_super) {
    __extends(SchemaMenuItem, _super);
    function SchemaMenuItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Create a new schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.create = function (body) {
        return this.rawCreate(Utils.serializeRequestBody(body, {
            type: 'schema_menu_item',
            attributes: ['label', 'position', 'kind'],
            relationships: ['item_type', 'parent'],
        })).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Create a new schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/create
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.rawCreate = function (body) {
        return this.client.request({
            method: 'POST',
            url: '/schema-menu-items',
            body: body,
        });
    };
    /**
     * Update a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.update = function (schemaMenuItemId, body) {
        return this.rawUpdate(Utils.toId(schemaMenuItemId), Utils.serializeRequestBody(body, {
            id: Utils.toId(schemaMenuItemId),
            type: 'schema_menu_item',
            attributes: ['label', 'position', 'kind'],
            relationships: ['item_type', 'parent', 'children'],
        })).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Update a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/update
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.rawUpdate = function (schemaMenuItemId, body) {
        return this.client.request({
            method: 'PUT',
            url: "/schema-menu-items/".concat(schemaMenuItemId),
            body: body,
        });
    };
    /**
     * List all schema menu items
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.list = function (queryParams) {
        return this.rawList(queryParams).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * List all schema menu items
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/instances
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.rawList = function (queryParams) {
        return this.client.request({
            method: 'GET',
            url: '/schema-menu-items',
            queryParams: queryParams,
        });
    };
    /**
     * Retrieve a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.find = function (schemaMenuItemId) {
        return this.rawFind(Utils.toId(schemaMenuItemId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Retrieve a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/self
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.rawFind = function (schemaMenuItemId) {
        return this.client.request({
            method: 'GET',
            url: "/schema-menu-items/".concat(schemaMenuItemId),
        });
    };
    /**
     * Delete a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.destroy = function (schemaMenuItemId) {
        return this.rawDestroy(Utils.toId(schemaMenuItemId)).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Delete a schema menu item
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/destroy
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     */
    SchemaMenuItem.prototype.rawDestroy = function (schemaMenuItemId) {
        return this.client.request({
            method: 'DELETE',
            url: "/schema-menu-items/".concat(schemaMenuItemId),
        });
    };
    /**
     * Reorders a set of schema menu items
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/reorder
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    SchemaMenuItem.prototype.reorder = function (body) {
        return this.rawReorder(Utils.serializeRequestBody(body, {
            type: 'schema_menu_item',
            attributes: ['position'],
            relationships: ['parent'],
        })).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Reorders a set of schema menu items
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/schema-menu_item/reorder
     *
     * @throws {ApiError}
     * @throws {TimeoutError}
     *
     * @deprecated This API call is to be considered private and might change without notice
     */
    SchemaMenuItem.prototype.rawReorder = function (body) {
        return this.client.request({
            method: 'POST',
            url: '/schema-menu-items/reorder',
            body: body,
        });
    };
    SchemaMenuItem.TYPE = 'schema_menu_item';
    return SchemaMenuItem;
}(BaseResource_1.default));
exports.default = SchemaMenuItem;
