// interactive-commander/dist/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  InteractiveCommand: () => InteractiveCommand,
  InteractiveOption: () => InteractiveOption
});
module.exports = __toCommonJS(src_exports);

// src/interactive-command.ts
var import_commander2 = require("commander");
var import_parse_my_command = require("parse-my-command");

// src/interactive-option.ts
var inquirer = __toESM(require("@inquirer/prompts"), 1);
var import_commander = require("commander");
var parseArgToValidate = (parseArg) => (value) => {
  if (!parseArg) {
    return true;
  }
  try {
    parseArg(value, void 0);
    return true;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    throw error;
  }
};
var InteractiveOption = class extends import_commander.Option {
  readFunction = this._defaultReadFunction;
  _promptFunctions;
  constructor(flags, description, promptFunctions) {
    super(flags, description);
    this._promptFunctions = promptFunctions ? {
      confirm: promptFunctions.confirm ?? inquirer.confirm,
      input: promptFunctions.input ?? inquirer.input,
      select: promptFunctions.select ?? inquirer.select
    } : inquirer;
  }
  /**
   * Set a function that will be called to read the option value interactively.
   * When undefined is passed, the prompt will be skipped.
   *
   * @param readFunction
   * @returns
   */
  prompt(readFunction) {
    this.readFunction = readFunction;
    return this;
  }
  async _defaultReadFunction(currentValue, option, _command) {
    if (option.isBoolean() || option.negate) {
      const answer2 = await this._promptFunctions.confirm({
        message: option.description,
        default: option.negate ? !currentValue : Boolean(currentValue)
      });
      return option.negate ? !answer2 : answer2;
    }
    if (option.argChoices) {
      const answer2 = await this._promptFunctions.select({
        message: option.description,
        choices: [...option.argChoices].sort(
          (value) => currentValue !== void 0 && value === currentValue || currentValue === void 0 && value === option.defaultValue ? -1 : 0
        ).map((choice) => ({ value: choice }))
      });
      return option.parseArg ? option.parseArg(answer2, void 0) : answer2;
    }
    const answer = await this._promptFunctions.input({
      message: option.description,
      default: currentValue,
      validate: parseArgToValidate(option.parseArg)
    });
    if (!answer || answer.length === 0) {
      return;
    }
    return option.parseArg ? option.parseArg(answer, void 0) : answer;
  }
};

// src/interactive-command.ts
var InteractiveCommand = class _InteractiveCommand extends import_commander2.Command {
  _interactiveOptionName = "interactive";
  /**
   * The command that its parseAsync method is called
   */
  _isRootCommand = false;
  _hasHook = false;
  // The following will only be set on the root command
  _providedOptions;
  _missingOptions;
  _providedOptionsSources;
  _pluginRegistrations = /* @__PURE__ */ new Set();
  createCommand(name) {
    return new _InteractiveCommand(name);
  }
  createOption(flags, description) {
    return new InteractiveOption(flags, description);
  }
  version(...args) {
    const returnValue = super.version(...args);
    const versionOptionName = this._versionOptionName;
    if (!versionOptionName) {
      return returnValue;
    }
    const versionOption = this.options.find(
      (option) => option.attributeName() === versionOptionName
    );
    versionOption.prompt(void 0);
    return returnValue;
  }
  /**
   * Enable interactive mode
   *
   * This method recursively auto-registers the "-i, --interactive" flag which
   * prompts the user for missing (interactive) options.
   *
   * You can optionally supply the flags and description to override the
   * defaults.
   *
   * This method should almost always be called on the root command, after all
   * subcommands are hooks are added/configured.
   */
  interactive(flags, description) {
    const newOption = new import_commander2.Option(
      flags ?? "-i, --interactive",
      description ?? "interactive mode"
    );
    const existingOption = this.options.find(
      (option) => option.attributeName() === newOption.attributeName()
    );
    if (!existingOption) {
      this.addOption(newOption);
      this._interactiveOptionName = newOption.attributeName();
    }
    this.addHookIfAbsent();
    for (const command of this.commands) {
      if (!(command instanceof _InteractiveCommand)) {
        continue;
      }
      command.interactive(flags, description);
    }
    return this;
  }
  parse(argv, options) {
    throw new Error("parse is not supported! Use parseAsync instead!");
  }
  async parseAsync(argv, options) {
    await Promise.all(this._pluginRegistrations);
    this._isRootCommand = true;
    try {
      const { providedOptions, missingOptions, providedOptionsSources } = (0, import_parse_my_command.partialParse)(this, argv ?? process.argv, options);
      this._providedOptions = providedOptions;
      this._missingOptions = missingOptions;
      this._providedOptionsSources = providedOptionsSources;
    } catch (error) {
      if (!(error instanceof import_commander2.CommanderError || // This is needed for the minified version
      error.name === "CommanderError")) {
        throw error;
      }
      const commanderError = error;
      if (!["commander.helpDisplayed"].includes(commanderError.code)) {
        super.error(commanderError.message, commanderError);
      }
    }
    return super.parseAsync(argv, options);
  }
  use(argument_) {
    const registerFunction = typeof argument_ === "string" ? async (command) => {
      const imported = await import(argument_);
      await imported.register(command);
    } : argument_;
    const result = Promise.resolve(registerFunction(this));
    this._pluginRegistrations.add(result);
    return this;
  }
  async readMissingOptions(options, providedOptions) {
    for (const option of options) {
      if (!(option instanceof InteractiveOption && option.readFunction)) {
        continue;
      }
      const key = option.attributeName();
      const value = await option.readFunction(
        this.getOptionValue(key) ?? providedOptions[key],
        option,
        this
      );
      if (option.required && value === void 0) {
        this.missingMandatoryOptionValue(option);
      }
      this.setOptionValue(option.attributeName(), value);
    }
  }
  findRootCommand() {
    let currentCommand = this;
    while (currentCommand) {
      if (currentCommand._isRootCommand) {
        return currentCommand;
      }
      currentCommand = currentCommand.parent;
    }
    throw new Error("Root command not found!");
  }
  addHookIfAbsent() {
    if (this._hasHook) {
      return;
    }
    this._hasHook = true;
    this.hook(
      "preSubcommand",
      async (thisCommand, actionCommand) => {
        if (!(thisCommand instanceof _InteractiveCommand) || !(actionCommand instanceof _InteractiveCommand)) {
          return;
        }
        if (!thisCommand.optsWithGlobals()[this._interactiveOptionName]) {
          return;
        }
        const rootCommand = this.findRootCommand();
        const providedOptions = rootCommand._providedOptions;
        const missingOptions = rootCommand._missingOptions;
        const providedOptionsSources = rootCommand._providedOptionsSources;
        for (const command of [thisCommand, actionCommand]) {
          const cliOptions = new Set(
            [...providedOptionsSources?.get(command)?.entries() ?? []].filter(([_, value]) => value === "cli").map(([key]) => key)
          );
          const optionKeys = [
            ...Object.keys(providedOptions?.get(command) ?? {}),
            ...missingOptions?.get(command)?.keys() ?? /* @__PURE__ */ new Set()
          ];
          const nonCliOptions = optionKeys.filter((key) => !cliOptions.has(key)).map(
            (key) => command.options.find((option) => option.attributeName() === key)
          ).filter(Boolean);
          nonCliOptions.sort((a, b) => {
            const aIndex = command.options.findIndex(
              (option) => option.attributeName() === a.attributeName()
            );
            const bIndex = command.options.findIndex(
              (option) => option.attributeName() === b.attributeName()
            );
            return aIndex - bIndex;
          });
          await command.readMissingOptions(
            nonCliOptions,
            providedOptions?.get(command) ?? {}
          );
        }
      }
    );
  }
};

// src/index.ts
__reExport(src_exports, require("commander"), module.exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  InteractiveCommand,
  InteractiveOption,
  ...require("commander")
});
